### ISSUE TITLE
Thread-safety in xarray

### ISSUE THREAD

**Issue by max-sixty** [pydata/xarray#9836]
### What is your issue?

There's a process that can't use xarray because it's a multithreaded context, and xarray objects aren't thread-safe, because

**Comment by shoyer** [comment]
My hope would be that pandas can fix this thread-safety issue upstream. Hopefully they are more amenable to doing so now (in 2025) than they were in 2013.

**Comment by alippai** [comment]
In a thread where `x1` and `x2` is shared I write:
```python
x3 = xr.merge(x1, x2)
```
why do I hit the above issue?
At least I get data duplication error randomly, so I guess I'm hitting this.
I assumed since I don't modify `x1` and `x2`, I shouldn't see threadsafety issues.

The actual message I get: `ValueError: cannot reindex or align along dimension '***' because the (pandas) index has duplicate values`

**Comment by max-sixty** [comment]
if that's the whole program, and there are no other threads, you shouldn't hit an issue

if that's running in a thread pool or similar, then you can get that sort of problem... (and iirc that is a message you would get...)

**Comment by alippai** [comment]
Yes, this is within threadpool.starmap, x1 and x2 passed as parameters

**Comment by max-sixty** [comment]
those sound like reasonable things to try! I'm not confident on the solution unfortunately

**Comment by mjwillson** [comment]
We have also seen this a lot even with read-only operations. It shows up a lot in data readers which are often run in a multithreaded context for prefetch etc.

I think **at least for immutable datatypes and/or read-only operations** pandas should really provide APIs which are thread-safe or have an option to be thread-safe. Sure, we may need extra coordination to prevent concurrent mutation of a shared mutable data structure. But lack of thread-safety for immutable objects and read-only ops just seems gratuitously difficult to me. IIRC at least part of the problem lies in things like memoized read optimizations that have been done in a non-threadsafe fashion in pandas.

While we wait for any movement from pandas, what could xarray do about it? Couple of preliminary ideas

* Lock around access to pandas.Index within xarray, with per-index granularity. May be a performance hit
* Make thread-local copies of pandas.Index instances (locking around the copy operation which is apparently non-threadsafe but should only need to happen once per index per thread).
  * Perhaps this could be done in a delegating wrapper around pandas.Index, and a make_threadsafe helper could wrap all the indexes of a Dataset in this wrapper, avoiding the need to touch xarray code as much?
  * Horrible hack but I wonder if `class ThreadlocalIndex(pandas.Index, threading.local):` would work, see https://docs.python.org/3/library/threading.html#thread-local-data.

**Comment by shoyer** [comment]
Strongly agreed! I am open to either of these, if it can done in a fashion
that is not too invasive into Xarray's codebase.
